objects/src/block/note_tree.rs

```diff
@@ -97,12 +91,6 @@ impl BlockNoteIndex {
         const _: () = assert!(MAX_NOTES_PER_BLOCK <= u32::MAX as usize);
         (self.batch_idx() * MAX_NOTES_PER_BATCH + self.note_idx_in_batch()) as u32
     }
-
-    /// Returns an index of the leaf containing the note.
-    fn leaf_index(&self) -> u32 {
-        const _: () = assert!(MAX_NOTES_PER_BLOCK * 2 <= u32::MAX as usize);
-        self.to_absolute_index() * 2
-    }
```

Comment:
New returning type is not easy to convert and it also requires to handle conversion errors. But let's see.

Reply:
The main inconvenience here that we often use leaf index as `node_index_in_block` argument of `BlockNoteIndex`, which is now `u16`. It was much easier, when both values were `u32`.

Reply:
It seems like at least here everything works more or less OK. Are the inconveniences in `miden-node`?

Reply:
Yes, all inconveniences which I faced are in `miden-node`. We have to construct `LeafIndex` first, construction might fail and we can't check this in compile-time, so we need to handle the result, then convert from its value, which is `u64`, to `u16` and also need to handle possible overflows. Sometimes we do it in iterator conversions and this require us to allocate temporary vector to process of conversion errors, instead of item mapping in a flow.

Reply:
Maybe we could also add `leaf_index_value()` method here that would look something like this:
```rust
pub fn leaf_index_value(&self) -> u16 {
    ...
}
```

Reply:
But with this solution we will have to do conversion under the hood and panic on conversion errors.

Reply:
But one example of iterators mapping which I don't like is here in `miden-base`: https://github.com/0xPolygonMiden/miden-base/pull/834/files#diff-0d4217a3e88df1dea33b5fa1477f9dec13390d254f30147fde9b66305fa9a021R39-R46

Reply:
> But one example of iterators mapping which I don't like is here in `miden-base`: https://github.com/0xPolygonMiden/miden-base/pull/834/files#diff-0d4217a3e88df1dea33b5fa1477f9dec13390d254f30147fde9b66305fa9a021R39-R46

Wouldn't this be just:
```rust
let leaves = entries
    .into_iter()
    .map(|(index, note_id, metadata)| {
        Ok((index.leaf_index_value() as u64, compute_note_hash(note_id, &metadata).into()))
    })
    .collect::<Result<Vec<_>, MerkleError>>()?;
```

> But with this solution we will have to do conversion under the hood and panic on conversion errors.

Do you mean inside `leaf_index_value()` function?

Reply:
Another option, we could probably implement `From<BlockNoteIndex> for u64` and then conversions should be pretty easy - no?

Reply:
> Do you mean inside `leaf_index_value()` function?

Yes. Please look into my current implementation of `leaf_index` method: https://github.com/0xPolygonMiden/miden-base/pull/834/files#diff-0d4217a3e88df1dea33b5fa1477f9dec13390d254f30147fde9b66305fa9a021R90-R94

The construction of `LeafIndex` might return error in a case of overflow (value higher than maximum value which corresponds to tree depth). Thus, if we implement `leaf_index_value`, we have to decide: return `Result<u64, Error>` or return just `u64` and panic in a case of overflow. Panic in libraries usually is not a good idea, so I'd return `Result` here. And this brakes iterator's mapping, we have to process results (re-throw error), so we have to allocate temporary vector. We can avoid additional allocation by making preliminary iteration over elements with checking conversion results and if everything is good, just `unwrap`/`expect` it in conversion flow. But this is also doesn't look good to me. 

Maybe I miss something and we can assume that the input of `BlockNoteIndex` is always correct or at least nearly impossible to overflow? In this case we can get rid of `Result`'s and just panic.

Reply:
Ah! I think we should make sure that `BlockNoteIndex` is always correct. That is, at construction time, we should make sure that `batch_idx` and `note_idx_in_batch` do not result in a note index over $2^{16}-1$.